### PR TITLE
Send CloseMessage in more cases

### PR DESCRIPTION
**Pull request by BrennanConroy**
The `CloseMessage(allowReconnect: false)` was not being sent when `Context.Abort()` was called, which resulted in clients potentially attempting a reconnect. This fixes it so that the `CloseMessage` can be sent now so well behaved clients wont attempt an unwanted reconnect.

Fixes https://github.com/dotnet/aspnetcore/issues/21422

### CODE REVIEW

**Comment by davidfowl** [comment]
Back port?

**Comment by BrennanConroy** [comment]
> Back port?

We can consider it once this is merged

File: src/SignalR/server/Core/src/HubConnectionHandler.cs [hunk 1]
```diff
@@ -226,7 +226,7 @@ private async Task SendCloseAsync(HubConnectionContext connection, Exception? ex
 
             try
             {
-                await connection.WriteAsync(closeMessage);
+                await connection.WriteCloseAsync(closeMessage);
```
Comment (halter73) [inline src/SignalR/server/Core/src/HubConnectionHandler.cs:229]: I prefer adding an `ignoreAbort = false` parameter to WriteAsync. Where it checks `if (_connectionAborted)` today, we'd change that to `if (_connectionAborted && !ignoreAbort)`. That way if we make a change to WriteAsync in the future, we get it for free without having to remember to also update WriteCloseAsync.

File: src/SignalR/server/Core/src/HubConnectionContext.cs [hunk 2]
```diff
@@ -146,6 +146,52 @@ internal StreamTracker StreamTracker
...
+        }
+
+        private async Task WriteCloseSlowAsync(CloseMessage message, CancellationToken cancellationToken)
+        {
+            // Failed to get the lock immediately when entering WriteAsync so await until it is available
+            await _writeLock.WaitAsync(cancellationToken);
+
+            try
+            {
+
```
Comment (halter73) [inline src/SignalR/server/Core/src/HubConnectionContext.cs:180]: Nit:

```suggestion
```

File: src/SignalR/server/Core/src/HubConnectionHandler.cs (hunk 1 above)
Comment (davidfowl) [reply src/SignalR/server/Core/src/HubConnectionHandler.cs]: and add a branch to every single write!? 😮

Comment (halter73) [reply src/SignalR/server/Core/src/HubConnectionHandler.cs]: But the branch is already there 😆

**Review by halter73 — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@Pilchie for RC2

**Comment by Pilchie** [comment]
Approved for .NET 5 RC2.

**Pull request merged**